### ISSUE TITLE
HTML streaming

### ISSUE THREAD

**Issue by gurgunday** [mcollina/fastify-html#11]
https://github.com/gurgunday/ghtml/commit/c06823f8df3875e917507c2585ac9ce0500e9ee8 added the `htmlGenerator` function, we can use it to stream templates as they get processed

I will probably send a PR in the following days, but just putting it up here so that if anyone is interested, they can give it a shot

**Comment by mcollina** [comment]
I don't think this is really useful, because generators are all synchronous anyway, and concatenating strings is incredibly cheap. I would expect generators to worsen performance.

However, supporting promises and streaming the chunk after each one resolves, would boost time to first byte.

**Comment by gurgunday** [comment]
> However, supporting promises and streaming the chunk after each one resolves, would boost time to first byte.

Alright, I will look into that, thanks for the input

**Comment by gurgunday** [comment]
Oh yeah, I get it now 😄

The sync generator achieved nothing I guess

Ass I said will take a look into the async version instead

**Comment by gurgunday** [comment]
Do you think the TTFB would be slower in the following scenarios? Even though the final result might take longer, the important thing here is that the user might potentially start seeing things much faster, no?

```js
app.get("/", function (req, res) {
  const htmlContent = html`<html>
    <p>${"...your HTML content..."}</p>
  </html>`;
  const readableStream = Readable.from(htmlContent);
  res.send(readableStream);
});
```

Or (of course will be adapted for Fastify):

```js
import { htmlGenerator as html } from "ghtml";
import http from "http";

http.createServer(async (req, res) => {
  if (req.url === '/stream') {
    res.writeHead(200, { 'Content-Type': 'text/html' });

  const htmlContent = html`<html>
  <p>${"...your HTML content..."}</p>
</html>`;
    for await (const chunk of htmlContent) {
      res.write(chunk);
    }

    res.end();
  } else {
    res.writeHead(404);
    res.end('Not Found');
  }
}).listen(3000, () => console.log('Server running on http://localhost:3000'));
```

**Comment by gurgunday** [comment]
Do you think the TTFB would be slower in the following scenarios? Even though the final result might take longer, the important thing here is that the user will start seeing things much faster, no?

```js
app.get("/", function (req, res) {
  const htmlContent = html`<html>
    <p>${"...your HTML content..."}</p>
  </html>`;
  const readableStream = Readable.from(htmlContent);
  res.send(readableStream);
});
```

Or (of course will be adapted for Fastify):

```js
import { htmlGenerator as html } from "ghtml";
import http from "http";

http.createServer(async (req, res) => {
  if (req.url === '/stream') {
    res.writeHead(200, { 'Content-Type': 'text/html' });

  const htmlContent = html`<html>
  <p>${"...your HTML content..."}</p>
</html>`;
    for await (const chunk of htmlContent) {
      res.write(chunk);
    }

    res.end();
  } else {
    res.writeHead(404);
    res.end('Not Found');
  }
}).listen(3000, () => console.log('Server running on http://localhost:3000'));
```

**Comment by gurgunday** [comment]
https://frontendmasters.com/blog/streaming-html/

**Comment by gurgunday** [comment]
Yeah I just watched [your masterclass](https://youtu.be/edB964-YYpE?si=7j0OL8PjBUPgC404), and to say the least, it came out at the perfect time 😄

So what I had in my mind with htmlGenerator won't improve TTFB, maybe I might add an async generator to combine with readFile, etc, but in any case thanks again for the information and the video 🙏

Closing for now

**Comment by mcollina** [comment]
It's slightly incorrect. If you have all the data, processing and rendering synchronously will be faster.

**Comment by gurgunday** [comment]
Yes, it says streaming with generators are **never** faster than processing directly without a generator

I'll try to improve the wording 😄

**Comment by gurgunday** [comment]
I updated it, it's clearer: https://github.com/gurgunday/ghtml/commit/210faf1f0f2f27c978c2a72af95f896e7bad610b#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R21-R23

**Comment by gurgunday** [comment]
Done: https://github.com/gurgunday/ghtml?tab=readme-ov-file#htmlgenerator
